hs_core/views/resource_folder_hierarchy.py

```diff
@@ -0,0 +1,313 @@
+import json
+import logging
+import os
+
+from django.http import HttpResponse
+from rest_framework.exceptions import NotFound, PermissionDenied
+
+from django_irods.storage import IrodsStorage
+from django_irods.icommands import SessionException
+
+from hs_core.hydroshare.utils import get_file_mime_type, get_resource_file_name_and_extension, \
+    get_resource_file_url
+from hs_core.views.utils import authorize, ACTION_TO_AUTHORIZE, zip_folder, unzip_file, \
+    create_folder, remove_folder, remove_or_rename_file_or_folder
+from hs_core.models import ResourceFile
+
+logger = logging.getLogger(__name__)
+
+
+def data_store_structure(request):
+    """
+    Get file hierarchy (collection of subcollections and data objects) for the requested directory
+    in hydroshareZone or any federated zone used for HydroShare resource backend store.
+    It is invoked by an AJAX call and returns json object that holds content for files
+    and folders under the requested directory/collection/subcollection.
+    The AJAX request must be a POST request with input data passed in for res_id and store_path
+    where store_path is the relative path under res_id collection/directory
+    """
+    res_id = request.POST['res_id']
+    if res_id is None:
+        return HttpResponse(status=400)
+    res_id = str(res_id).strip()
+    try:
+        resource, _, _ = authorize(request, res_id,
+                                   needed_permission=ACTION_TO_AUTHORIZE.VIEW_RESOURCE)
+    except (NotFound, PermissionDenied):
+        # return permission defined response
+        return HttpResponse(status=403)
+
+    store_path = request.POST['store_path']
+    if store_path is None:
+        return HttpResponse(status=400)
+    store_path = str(store_path).strip()
+    if not store_path:
+        return HttpResponse(status=400)
+    if resource.resource_federation_path:
+        istorage = IrodsStorage('federated')
+        res_coll = os.path.join(resource.resource_federation_path, res_id, store_path)
+    else:
+        istorage = IrodsStorage()
+        res_coll = os.path.join(res_id, store_path)
+
+    try:
+        store = istorage.listdir(res_coll)
+        files = []
+        for fname in store[1]:
+            name_with_full_path = os.path.join(res_coll, fname)
+            size = istorage.size(name_with_full_path)
+            mtype = get_file_mime_type(fname)
+            idx = mtype.find('/')
+            if idx >= 0:
+                mtype = mtype[idx + 1:]
+            f_pk = ''
+            f_url = ''
+            for f in ResourceFile.objects.filter(object_id=resource.id):
+                if fname == get_resource_file_name_and_extension(f)[1]:
+                    f_pk = f.pk
+                    f_url = get_resource_file_url(f)
+                    break
+            files.append({'name': fname, 'size': size, 'type': mtype, 'pk': f_pk, 'url': f_url})

```

Comment:
The front-end does not check file extension but uses the mime type returned from back-end directly, which is why I added mimetype as part of file attribute to return to front-end here.